app/services/kessel_rbac.rb

```diff
@@ -0,0 +1,260 @@
+        raise AuthorizationError, "Workspace listing failed: #{e.message}"
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    delegate :get_default_workspace_id, to: :KesselUtils
+
+    private
+
+    def build_client
+      raise ConfigurationError, 'Kessel is not enabled' unless enabled?
+
+      # Create gRPC client using the actual Kessel SDK API
+      if Settings.kessel.insecure
+        build_insecure_client
+      else
+        build_secure_client
+      end
+    rescue StandardError => e
+      Rails.logger.error("Failed to build Kessel client: #{e.message}")
+      raise ConfigurationError, "Failed to build Kessel client: #{e.message}"
+    end
+
+    def build_insecure_client
+      KesselInventoryService::ClientBuilder.new(Settings.kessel.url).insecure.build
+    end
+
+    def build_secure_client
+      builder = KesselInventoryService::ClientBuilder.new(Settings.kessel.url)
+
+      if Settings.kessel.auth.enabled
+        builder.oauth2_client_authenticated(auth)
+      else
+        builder.authenticated
+      end
+
+      builder.build
+    end
+
+    # rubocop:disable Metrics/AbcSize
+    def build_oauth_credentials
+      # Fetch OIDC discovery metadata
+      discovery = fetch_oidc_discovery(Settings.kessel.auth.oidc_issuer)
+
+      # Create OAuth2 client credentials
+      OAuth2ClientCredentials.new(
+        client_id: Settings.kessel.auth.client_id,
+        client_secret: Settings.kessel.auth.client_secret,
+        token_endpoint: discovery.token_endpoint
+      )
+    rescue StandardError => e
+      Rails.logger.error("Failed to build OAuth credentials: #{e.message}")
+      raise ConfigurationError, "Failed to build OAuth credentials: #{e.message}"
+    end
+    # rubocop:enable Metrics/AbcSize
+
+    def build_resource_reference(resource_type, resource_id, reporter_type)
+      ResourceReference.new(
+        resource_type: resource_type,
+        resource_id: resource_id,
+        reporter: ReporterReference.new(type: reporter_type)
+      )
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def build_subject_reference(user)
+      identity = user.account.identity_header.identity
+
+      identity_type = identity&.dig('type')
+
+      principal_id = if identity_type == 'ServiceAccount'
+                       identity&.dig('service_account', 'user_id')
+                     elsif identity_type == 'User'
+                       identity&.dig('user', 'user_id')
+                     else
+                       raise 'unsupported identity type'
+                     end
+
+      SubjectReference.new(
+        resource: ResourceReference.new(
+          resource_type: 'principal',
+          resource_id: "redhat/#{principal_id}",
+          reporter: ReporterReference.new(type: 'rbac')
+        )
+      )
+    end
+    # rubocop:enable Metrics/MethodLength
+  end
+end
+
+# rubocop:enable Metrics/ClassLength
+# Utilities used by Kessel
+class KesselUtils
```

Comment:
Utils could be extracted into a separate file

Reply:
Updated